trin-utils/src/db.rs

```diff
@@ -0,0 +1,236 @@
+use std::os::unix::fs::symlink;
+use std::path::PathBuf;
+use std::{env, fs};
+
+use anyhow::{anyhow, bail};
+use directories::ProjectDirs;
+use discv5::enr::{CombinedKey, EnrBuilder, NodeId};
+use ethereum_types::H256;
+use tempfile::TempDir;
+use tracing::debug;
+
+use crate::bytes::{hex_decode, hex_encode};
+
+const TRIN_DATA_ENV_VAR: &str = "TRIN_DATA_PATH";
+const TRIN_DATA_DIR: &str = "trin";
+const ACTIVE_NODE_DATA_DIR: &str = "active_node_data_dir.txt";
+const UNSAFE_PRIVATE_KEY_FILE_NAME: &str = "unsafe_private_key.txt";
+
+/// Create a directory on the file system that is deleted once it goes out of scope
+fn setup_temp_dir() -> anyhow::Result<TempDir> {
+    let mut os_temp = env::temp_dir();
+    os_temp.push(TRIN_DATA_DIR);
+    debug!("Creating temp dir: {os_temp:?}");
+    fs::create_dir_all(&os_temp)?;
+
+    let temp_dir = TempDir::new_in(&os_temp)?;
+    env::set_var(TRIN_DATA_ENV_VAR, temp_dir.path());
+
+    Ok(temp_dir)
+}
+
+/// Return the default trin data directory
+fn trin_data_dir() -> anyhow::Result<PathBuf> {
+    let trin_data_dir = match env::var(TRIN_DATA_ENV_VAR) {
+        Ok(val) => val,
+        Err(_) => get_default_data_dir()?,
+    };
+    Ok(PathBuf::from(&trin_data_dir))
+}
+
+/// Returns the data directory associated with the provided node id. Does not
+/// validate that the node id is active. Returns an error if the data dir does not exist, unless we
+/// are in ephemeral mode.
+pub fn get_node_data_dir(node_id: NodeId) -> anyhow::Result<PathBuf> {
+    let trin_data_dir = trin_data_dir()?;
+    if !trin_data_dir.exists() {
+        bail!(
+            "Trin data dir does not exist. Expected to find valid data dir at: {trin_data_dir:?}"
+        );
+    }
+    let application_string = get_data_dir_name_from_node_id(node_id)?;
+    let node_data_dir = trin_data_dir.join(application_string);
+    // this exception is for ephemeral mode to prevent the construction of excessive tmpdirs
+    // when running multiple nodes in peertest
+    if !node_data_dir.exists() && !node_data_dir.to_str().unwrap().contains("tmp") {
```

Comment:
Yeah, I can't say I love this but "it gets the tests passing". Though, that's a poor justification...

From the best of what I can tell, when we run peertest, we launch 3 ephemeral nodes. Each one of these creates a new temp dir on startup. The node data dir gets created inside each respective temp dir. **But the temp dir env variable is set on each startup**.

So, for example when node 1 goes to lookup it's node data dir, it tries to look it up inside of node 3's temp trin dir, which throws an error, causing peertest to fail.

The behavior that I changed in this pr which causes this is to return an error for a failed node data dir lookup. Previously we would just create the data dir if it was not found. I'm not strongly attached to this new behavior, but I do think it's beneficial to not blindly create node data dirs, if the lookup failed.

I'm not convinced that this "hack" is the best solution. Maybe introducing a `--trin-data-dir` cli arg rather than using an env variable? But, the only scenario (outside of peertest) that I can think of when we'll be using multiple ephemeral instances is in bridge mode, but then we're operating with `--kb=0` so there shouldn't be any tangible consequences.

Reply:
Well, maybe scratch everything that I wrote above... The hack led to some other testing errors. But I'll leave the comment there for reference until I sort things out.

Reply:
Ok, was able to remove the hack. Instead of running each node during peertest in `--ephemeral` mode, we now just setup a temp dir as `TRIN_DATA_DIR` at the beginning of the peertest.

Reply:
Okay, still a *bit* of a hack, but a much more tolerable one to me. :)

I've got what I think is the "proper" solution in my previous review comment, but it's up to you.